src/linuxkernel.rs

```diff
@@ -86,23 +86,54 @@ impl<'a, R: Read + Seek> LinuxKernel<'a, R> {
 
 impl<'a, R: Read + Seek> VersionFinder for LinuxKernel<'a, R> {
     fn get_version(&mut self) -> Option<String> {
-        let buffer = match discover_linux_kernel_kind(self.buf)? {
+        match discover_linux_kernel_kind(self.buf)? {
             LinuxKernelKind::ARMzImage => {
-                let mut buffer = [0; 0x200];
-                let buf: &mut [u8] = &mut buffer;
-
-                // Read the Linux kernel version from the reader
-                match compress_tools::uncompress_data(&mut self.buf, buf) {
-                    Ok(_) => {}
-                    // Write will most likely overflow the 0x200 bytes slice,
-                    // which is ok sincewe don't need more than that.
-                    Err(compress_tools::Error::Io(ref e))
-                        if e.kind() == io::ErrorKind::WriteZero => {}
-
-                    Err(_) => return None,
+                fn get_version_from_arm<R: Read>(mut rd: R) -> Option<String> {
+                    let mut buffer = Vec::default();
+                    compress_tools::uncompress_data(&mut rd, &mut buffer).ok()?;
+                    let re = Regex::new(r"Linux version (?P<version>\S+).*").unwrap();
+                    let m = re.captures(&buffer)?;
+                    let version = m.name("version")?.as_bytes().to_vec();
+                    Some(String::from_utf8(version).ok()?)
                 }
 
-                buffer
+                let mut buffer = [0; 0x200];
+                loop {
+                    let n = self.buf.read(&mut buffer).ok()?;
+                    if n == 0 {
+                        // EOF
+                        return None;
+                    }
+
+                    // Look for compression format header
+                    for (offset, window) in buffer[0..n].windows(6).enumerate() {
+                        // Headers taken from:
+                        // https://github.com/torvalds/linux/blob/master/scripts/extract-vmlinux
+                        match window {
+                            [0x1f, 0x8b, 0x08, ..] => {}               // gzip
+                            [0xfd, b'7', b'z', b'X', b'Z', 0x00] => {} // xz
+                            [b'B', b'Z', b'h', ..] => {}               // bzip2
+                            [0x5d, 0x00, 0x00, ..] => {}               // lzma
+                            [0x89, 0x4c, 0x5a, ..] => {}               // lzo
+                            [0x02, b'!', b'L', 0x18, ..] => {}         // lz4
+                            [b'(', 0xb5, b'/', 0xfd, ..] => {}         // zstd
+                            _ => continue,
+                        }
+
+                        let mut slice = &buffer[offset..];
+                        let current = self.buf.seek(SeekFrom::Current(0)).ok()?;
+                        let rd = io::Read::chain(&mut slice, &mut self.buf);
```

Comment:
`&mut buffer[offset..]` drops an indirection.

Reply:
Not possible, since with `&mut buffer[offset..]` we have that `[u8]` itself is not `Read`; and `&mut &buffer[offset..]` would not meet lifetime requirements.